crates/trie/trie/src/proof_v2/node.rs

```diff
@@ -0,0 +1,157 @@
+use crate::proof_v2::DeferredValueEncoder;
+use alloy_rlp::Encodable;
+use alloy_trie::nodes::ExtensionNodeRef;
+use reth_execution_errors::trie::StateProofError;
+use reth_trie_common::{
+    BranchNode, ExtensionNode, LeafNode, LeafNodeRef, Nibbles, RlpNode, TrieMask, TrieNode,
+};
+
+/// A trie node which is the child of a branch in the trie.
+#[derive(Debug)]
+pub(crate) enum ProofTrieBranchChild<RF> {
+    /// A leaf node whose value has yet to be calculated and encoded.
+    Leaf {
+        /// The short key of the leaf.
+        short_key: Nibbles,
+        /// The [`DeferredValueEncoder`] which will encode the leaf's value.
+        value: RF,
+    },
+    /// An extension node whose child branch has not yet been converted to an [`RlpNode`]
+    Extension {
+        /// The short key of the leaf.
+        short_key: Nibbles,
+        /// The node of the child branch.
+        child: BranchNode,
```

Comment:
I'm still not 100% this has to be BranchNode, vs just being the `RlpNode` of the branch. Will know for sure after the next PR